### PR TITLE
Allow map memory drawing mode selection through options

### PR DESCRIPTION
**Pull request by ZhilkinSerg**
#### Summary

```SUMMARY: Interface "Allow map memory drawing mode selection through options"```

#### Purpose of change

Allow map memory drawing mode selection through options

#### Describe the solution

Exposed map memory mode selection to Graphics tab of Options menu. Defaults to "Modern (Sepia)" and also allows selection of "Classic (Gray)" value. Changing this options requires restart or tileset reload.

#### Describe alternatives you've considered

As outlined in #32669 we want to make this option configurable (either in a tileset or via a tileset mod). Current PR is just a band-aid for those retrogrades who don't like sepia.

#### Additional context

image

### CODE REVIEW

**Comment by MeltedClowns** [comment]
This is exactly what I want the second the update came out! (Glad I'm not alone)

**Comment by ZhilkinSerg** [comment]
Jenkins rebuild

**Comment by ProfoundDarkness** [comment]
Count me among the anxious for this as well.

**Comment by ZhilkinSerg** [comment]
Jenkins rebuild.

**Comment by kevingranade** [comment]
This pull request has been mentioned on **Cataclysm: Dark Days Ahead**. There might be relevant details there:

https://discourse.cataclysmdda.org/t/going-back-to-original-grey-instead-of-sepia-for-memory/20940/2

**Comment by kasanryukin** [comment]
Sepia looks horrible on my monitor, and I'm willing to bet it does for a lot of folks too.  I'm glad to see a quick fix to have a choice between classic grey and modern sepia is ready to be merged.

**Comment by kevingranade** [comment]
This pull request has been mentioned on **Cataclysm: Dark Days Ahead**. There might be relevant details there:

https://discourse.cataclysmdda.org/t/how-to-change-sepia-brown-to-old-gray-colors-tileset-only-compile-only/20948/2

**Comment by esotericist** [comment]
I think calling this "gray" or "grayscale" is misleading; we use grayscale for low-light vision, and it's genuine grayscale, while the "classic" map memory was full-color but reduced in intensity (and isn't actually having its saturation reduced, so it's no more "gray" than the default tiles, just darker).

Maybe use different terminology? "Muted" is the best I can come up with offhand, "darkened" is another option.

**Comment by chphilli** [comment]
I'm curious about the status of this PR. I suspect that there's a performance issue with the sepia feature at least on my machine when moving around causes bunch of tiles to need to be drawn in sepia -- and based on my reading of the code this update will allow me to test the paths. 

I don't quite have my setup in place to build it myself and test it, so I'm hoping that this PR goes through and I can test that way. 

Thanks!

**Comment by esotericist** [comment]
> I'm curious about the status of this PR. I suspect that there's a performance issue with the sepia feature at least on my machine when moving around causes bunch of tiles to need to be drawn in sepia -- and based on my reading of the code this update will allow me to test the paths.
> 
> I don't quite have my setup in place to build it myself and test it, so I'm hoping that this PR goes through and I can test that way.
> 
> Thanks!

The sepia change only altered how the memory map tile assets are generated. This only occurs when the tileset is reloaded. 

It should not be possible for the sepia change to impact runtime performance.

**Comment by ZhilkinSerg** [comment]
Status of the PR is open.

There are no performance in-game issues either with sepia or without it. All tile variants are created during tileset loading - usually once per game session.

File: src/sdl_utils.h [hunk 1]
```diff
@@ -3,10 +3,16 @@
 #define SDL_UTILS_H
 
 #include <algorithm>
+#include <unordered_map>
```
Comment (ZhilkinSerg) [inline src/sdl_utils.h]:
```suggestion
#include <cmath>
#include <unordered_map>
```

File: src/options.cpp [hunk 2]
```diff
@@ -1674,6 +1674,17 @@ void options_manager::add_options_graphics()
 
     get_option( "TILES" ).setPrerequisite( "USE_TILES" );
 
+    mOptionsSort["graphics"]++;
+
+    add( "MEMORY_MAP_MODE", "graphics", translate_marker( "Memory map drawing mode" ),
+    translate_marker( "Specified the mode in which the memory map is drawn.  Requires restart." ), {
+        { "color_pixel_gray_out", translate_marker( "Classic (Gray)" ) },
```
Comment (ZhilkinSerg) [inline src/options.cpp]:
```suggestion
        { "color_pixel_gray_out", translate_marker( "Classic (Darken)" ) },
```

File: src/sdl_utils.cpp [hunk 3]
```diff
@@ -10,6 +10,25 @@
 #include "cursesport.h"
 #include "sdltiles.h"
 
+color_pixel_function_map builtin_color_pixel_functions = {
+    { "color_pixel_none", nullptr },
+    { "color_pixel_gray_out", color_pixel_gray_out },
```
Comment (ZhilkinSerg) [inline src/sdl_utils.cpp]:
```suggestion
    { "color_pixel_darken", color_pixel_darken },
```

File: src/sdl_utils.h [hunk 4]
```diff
@@ -96,7 +102,23 @@ inline SDL_Color color_pixel_overexposed( const SDL_Color &color )
     };
 }
 
-inline SDL_Color color_pixel_memorized( const SDL_Color &color )
+inline SDL_Color color_pixel_gray_out( const SDL_Color &color )
```
Comment (ZhilkinSerg) [inline src/sdl_utils.h]:
```suggestion
inline SDL_Color color_pixel_darken( const SDL_Color &color )
```

File: src/options.cpp [hunk 5]
```diff
@@ -1671,6 +1671,17 @@ void options_manager::add_options_graphics()
 
     get_option( "TILES" ).setPrerequisite( "USE_TILES" );
 
+    mOptionsSort["graphics"]++;
+
+    add( "MEMORY_MAP_MODE", "graphics", translate_marker( "Memory map drawing mode" ),
+    translate_marker( "Specified the mode in which the memory map is drawn.  Requires restart." ), {
+        { "color_pixel_darken", translate_marker( "Classic (Darkened)" ) },
+        { "color_pixel_sepia", translate_marker( "Modern (Sepia)" ) }
```
Comment (Night-Pryanik) [inline src/options.cpp]: I don't think both "Classic" and "Modern" names are really needed here. What if someone adds another drawing mode? Call it "Super-Modern" or "Neo-Modern"? Simple "Darkened" and "Sepia" is enough.

Comment (ZhilkinSerg) [inline src/options.cpp]:
```suggestion
        { "color_pixel_sepia", translate_marker( "Sepia" ) }
```

File: src/options.cpp [hunk 6]
```diff
@@ -1671,6 +1671,17 @@ void options_manager::add_options_graphics()
 
     get_option( "TILES" ).setPrerequisite( "USE_TILES" );
 
+    mOptionsSort["graphics"]++;
+
+    add( "MEMORY_MAP_MODE", "graphics", translate_marker( "Memory map drawing mode" ),
+    translate_marker( "Specified the mode in which the memory map is drawn.  Requires restart." ), {
+        { "color_pixel_darken", translate_marker( "Classic (Darkened)" ) },
```
Comment (ZhilkinSerg) [inline src/options.cpp]:
```suggestion
        { "color_pixel_darken", translate_marker( "Darkened" ) },
```

**Comment by esotericist** [comment]
Well, I'm happy with this.

**Comment by chronicpayne** [comment]
NUyttbn

**Pull request merged**